gadget/update.go

```diff
@@ -21,12 +21,78 @@ package gadget
 import (
 	"errors"
 	"fmt"
+
+	"github.com/snapcore/snapd/logger"
 )
 
 var (
-	ErrNoUpdate = errors.New("no update needed")
+	ErrNoUpdate = errors.New("nothing to update")
+
+	// default positioning constraints that match ubuntu-image
+	defaultConstraints = PositioningConstraints{
+		NonMBRStartOffset: 1 * SizeMiB,
+		SectorSize:        512,
+	}
 )
 
+// GadgetData holds references to a gadget revision metadata and its data directory.
+type GadgetData struct {
+	// Info is the gadget metadata
+	Info *Info
+	// RootDir is the root directory of gadget snap data
+	RootDir string
+}
+
+// Update applies the gadget update given the gadget information and data from
+// old and new revisions. It errors out when update is not possible or illegal,
+// or a failure occurs at any of the steps. When there is no update, a special
+// error ErrNoUpdate is returned.
+//
+// Updates are opt-in, and are only applied to structures which have a different
+// value of Edition field between old and new gadget definition.
+//
+// Data that would be modified during the update is first backed up inside the
+// rollback directory. Should the apply step fail, the modified data is
+// recovered.
+func Update(old, new GadgetData, rollbackDirPath string) error {
+	oldVol, newVol, err := resolveVolume(old.Info, new.Info)
+	if err != nil {
+		return err
+	}
+
+	// layout old
+	pOld, err := PositionVolume(old.RootDir, oldVol, defaultConstraints)
```

Comment:
sorry for noticing this only now,  but "lay out" would probably be a more typical verb for this?  did position come from the ubuntu-image code?

Reply:
As discussed yesterday, we'll update the naming to reflect the fact that it's about volume/partition layout. However, so as to keep the conflicts limited, this will happen once more of the pending PRs land.